src/catalog/dependency_manager.cpp

```diff
@@ -128,26 +129,110 @@ void DependencyManager::EraseObject(CatalogEntry &object) {
-		if (entry != dependents_map.end()) {
-			D_ASSERT(entry->second.find(object) != entry->second.end());
-			entry->second.erase(object);
+
+	auto &entries = GetEntriesThatObjectDependsOn(object);
+	for (auto &other : entries) {
+		// For every entry that 'object' is dependent on, clean up this connection
+		auto dependencies_entry = dependents_map.find(other);
+		if (dependencies_entry == dependents_map.end()) {
+			continue;
 		}
+
+		auto &dependencies = dependencies_entry->second;
+		auto dependent_entry = dependencies.find(object);
+		D_ASSERT(dependent_entry != dependencies.end());
+
+		// Remove the dependency of 'object' on 'other'
+		dependencies.erase(dependent_entry);
 	}
 	// erase the dependents and dependencies for this object
 	dependents_map.erase(object);
 	dependencies_map.erase(object);
 }
 
+void DependencyManager::PrintDependencyMap() {
+	Printer::Print("DEPENDENCIES_MAP");
+	for (auto &entry : dependencies_map) {
+		Printer::Print(entry.first.get().ToSQL());
+		for (auto &other : entry.second) {
+			Printer::Print("\t" + other.get().ToSQL());
+		}
+	}
+}
+
+void DependencyManager::PrintDependentsMap() {
+	Printer::Print("DEPENDENTS_MAP");
+	for (auto &entry : dependents_map) {
+		Printer::Print(entry.first.get().ToSQL());
+		for (auto &other : entry.second) {
+			Printer::Print("\t" + other.entry.get().ToSQL());
+		}
+	}
+}
+
+bool DependencyManager::AllExportDependenciesWritten(CatalogEntry &object, catalog_entry_set_t &dependencies,
+                                                     catalog_entry_set_t &exported) {
+	for (auto &entry : dependencies) {
+		// This is an entry that needs to be written before 'object' can be written
+		if (exported.find(entry) == exported.end()) {
+			// It has not been written yet, abort
+			return false;
+		}
+		// We do not need to check recursively, if the object is written
+		// that means that the objects it depends on have also been written
+	}
+	return true;
+}
+
+catalog_entry_vector_t DependencyManager::GetExportOrder() {
+	catalog_entry_set_t entries;
+	catalog_entry_vector_t export_order;
+
+	queue<reference<CatalogEntry>> backlog;
+	// Populate the backlog with every entry in the dependencies map
+	for (auto &entry : dependencies_map) {
+		backlog.push(entry.first);
+	}
+
+	// First populate our backlog with every entry in dependencies_map
+	while (!backlog.empty()) {
+		auto &object = backlog.front();
+		backlog.pop();
+		if (entries.count(object)) {
+			// This entry has already been written
+			continue;
+		}
+		auto entry = dependencies_map.find(object);
+		if (entry == dependencies_map.end() || AllExportDependenciesWritten(object, entry->second, entries)) {
+			// All dependencies written, we can write this now
+			auto insert_result = entries.insert(object);
+			D_ASSERT(insert_result.second);
+			export_order.push_back(object);
+		} else {
+			for (auto &dependency : entry->second) {
+				backlog.emplace(dependency);
+			}
+			backlog.emplace(object);
+		}
+	}
```

Comment:
This implementation is potentially O(n^2), not sure we care enough and it's not immediate to me what an alternative implementation would be.

Reply:
If we turn this into a `stack` instead and push every dependency to the front, this should be slightly more efficient, as we won't keep checking the same object if the dependencies of the object also have dependencies

now:
obj has dependencies:

state of the queue:
```
dep1
dep2
dep3
obj
```

dep1 has dependencies, push these, state of the queue:
```
dep2
dep3
obj
dep1.1
dep1.2
dep1.3
dep1
```
Once we reach `obj` again, we will push `dep1` and the rest of its dependencies again (dep2, dep3)

With a stack:
obj has dependencies:
```
dep3
dep2
dep1
obj
```

dep1 has dependencies, push these:
```
dep1.3
dep1.2
dep1.1
dep1
obj
```

Reply:
to get the order don't you just do a depth first search over the entries added to the backlog? my c++ is rusty but something like:

```
oid DependencyManager::DepthFirstSearch(catalog_entry_set_t &explored, catalog_entry_vector_t &order, the_type &entry)
{
        if (explored.count(entry) > 0) {
                return;
        }
        explored.insert(entry);
        auto deps = dependencies_map.find(entry);
        if (deps != dependencies_map.end() {
                for (auto &dependency : deps->second) {
                        DepthFirstSearch(explored, order, dependency);
                }
        }   
        order.push_back(entry);
}

catalog_entry_vector_t DependencyManager::GetExportOrder() {
        catalog_entry_set_t explored;
        catalog_entry_vector_t export_order;

        for (auto &entry : dependencies_map) {
                DepthFirstSearch(explored, export_order, entry.first);
        }
        return export_order;
}
```

Reply:
I wanted to avoid recursion, to not run the risk of a stack overflow.
But maybe I was being too cautious for no good reason here

Reply:
i don't quite follow the explanation with the stack / queue described. For a non-recursive DFS there is an implementation on Wikipedia, https://en.wikipedia.org/wiki/Depth-first_search#Pseudocode, that uses a stack of iterators.  That would work to give a topological sort -- adding the element to the output stack at the last pop in the pseudo-code.

Reply:
here's my attempt, i was bored.

https://github.com/robert-manchester/duckdb/blob/3dd41cf54918aa722fabecb6a45137e6b88c28f5/src/catalog/dependency_manager.cpp#L189